{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAGCATAAGGGC\n",
      "GTTTAGGTACAA\n",
      "GGTTCCTACTCC\n",
      "CGCGGACGCATG\n",
      "GGTTATGAAATG\n",
      "GGCCGAGAATAC\n",
      "GACTACTTCTCG\n",
      "GGCTAATGGACC\n",
      "GATTGCTGAACG\n",
      "CGTTACGACAAG\n",
      "GATTACGTGTGC\n",
      "GTGTGCTACTAG\n",
      "GGCCACTTGTGC\n",
      "GATTGCCTCTAC\n",
      "GACTACCTGTGC\n",
      "GTCTACCTGTGC\n",
      "GATGACGTGTGC\n",
      "CTTTGCCGGTAG\n",
      "CTGTGCTAGTGC\n",
      "GTGTGCGTATCC\n",
      "GACGACTTGTGC\n",
      "GGGCACTTGTGC\n",
      "GAGTACGTGTGC\n",
      "GTTCACATGTGC\n",
      "GGGCACCTGTGC\n"
     ]
    }
   ],
   "source": [
    "def read_data_from(filename):\n",
    "    with open(filename) as f:\n",
    "        data = f.readlines()\n",
    "\n",
    "    for i in range(0, len(data)):\n",
    "        data[i] = data[i].strip()\n",
    "\n",
    "    return data\n",
    "\n",
    "profile_row = {'A': 0, 'C':1, 'G':2, 'T':3}\n",
    "def make_profile(kMers):\n",
    "    \"\"\" return list of dictionaries where each dictionary is column of the profile \"\"\"\n",
    "\n",
    "    if len(kMers) == 0:\n",
    "        print(\"\"\"None\"\"\")\n",
    "        return None\n",
    "\n",
    "    k = len(kMers[0])\n",
    "    t = len(kMers)\n",
    "    profile = list(dict())\n",
    "    for i in range(k):\n",
    "        profile_col = {'A':0, 'C':0, 'G':0, 'T':0}\n",
    "        for j in range(t):\n",
    "            profile_col[kMers[j][i]] += 1\n",
    "\n",
    "        for key in profile_col.keys():\n",
    "            profile_col[key] = 1.0 * profile_col[key] / t\n",
    "\n",
    "        profile.append(profile_col)\n",
    "\n",
    "    return profile\n",
    "\n",
    "def pattern_probability(profile, pattern):\n",
    "    probability = 1\n",
    "    for i in range(0, len(pattern)):\n",
    "        probability *= profile[i][pattern[i]]\n",
    "\n",
    "    return probability\n",
    "\n",
    "def profile_most_probable_kmer(dna, profile, k):\n",
    "    start = 0\n",
    "    length = len(dna)\n",
    "    max_probability = 0\n",
    "    most_probable = dna[0:k]\n",
    "    while start + k <= length:\n",
    "        substr = dna[start:start+k]\n",
    "        probability = pattern_probability(profile, substr)\n",
    "        if probability > max_probability:\n",
    "            most_probable = substr\n",
    "            max_probability = probability\n",
    "\n",
    "        start += 1\n",
    "\n",
    "    return most_probable\n",
    "\n",
    "def hamming_dist(str_one, str_two):\n",
    "    \"\"\" returns number of hamming_dist between two strings \"\"\"\n",
    "\n",
    "    len_one = len(str_one)\n",
    "    len_two = len(str_two)\n",
    "    if len_one != len_two:\n",
    "        raise ValueError(\"Strings have different lengths.\")\n",
    "\n",
    "    mismatches = 0\n",
    "    for i in range(len_one):\n",
    "        if str_one[i] != str_two[i]:\n",
    "            mismatches += 1\n",
    "\n",
    "    return mismatches\n",
    "\n",
    "def make_consensus(motifes):\n",
    "    profile = make_profile(motifes)\n",
    "    consensusList = list()\n",
    "    for item in profile:\n",
    "        consensusList.append(max(item, key=item.get))\n",
    "\n",
    "    return ''.join(consensusList)\n",
    "\n",
    "def score(motifes):\n",
    "    consensus = make_consensus(motifes)\n",
    "    score = 0\n",
    "    for motif in motifes:\n",
    "        score += hamming_dist(consensus, motif)\n",
    "\n",
    "    return score\n",
    "\n",
    "def greedy_motif_search(Dna, k, t):\n",
    "    if len(Dna) == 0:\n",
    "        print (\"None!!!\")\n",
    "        return None\n",
    "\n",
    "    bestMotif = list()\n",
    "    for i in range(t):\n",
    "        bestMotif.append(Dna[i][0:k])\n",
    "\n",
    "    start = 0\n",
    "    length = len(Dna[0])\n",
    "    while start + k <= length:\n",
    "        motifes = list()\n",
    "        motifes.append(Dna[0][start:start+k])\n",
    "        for i in range(1, t):\n",
    "            profile = make_profile(motifes[0:i])\n",
    "            motifes.append(profile_most_probable_kmer(Dna[i], profile, k))\n",
    "\n",
    "        if score(motifes) < score(bestMotif):\n",
    "            bestMotif = motifes[:]\n",
    "\n",
    "        start += 1\n",
    "\n",
    "    for element in bestMotif:\n",
    "        print (element)\n",
    "\n",
    "    return bestMotif\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    Dna = read_data_from(\"inputgreedy.txt\")\n",
    "    greedy_motif_search(Dna, 12, 25)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Greedy motif search algorithm is to find the set of motifs across a number of DNA sequences that match each other most closely.\n",
    "Greedy algorithms select the “most attractive” alternative at each iteration. \n",
    "GreedyMotifSearch, starts by forming a motif matrix from arbitrarily selected k-mers in each string from Dna (which in our specific implementation is the first k-mer in each string). It then attempts to improve this initial motif matrix by trying each of the k-mers in Dna1 as the first motif. \n",
    "GreedyMotifSearch constructs Profile(Motifs) and selects the Profile-most probable k-mer from Dnai based on this profile matrix. After obtaining a k-mer from each string to obtain a collection Motifs, GreedyMotifSearch tests to see whether Motifs outscores the current best scoring collection of motifs and then moves Motif1 one symbol over in Dna1, beginning the entire process of generating Motifs again."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First we have function to read file and then we have profile to set up score for each nucleotide  {'A': 0, 'C':1, 'G':2, 'T':3}\n",
    "Then we have function to make profile where we have love that we set all the score for each nucleotide are 0 at first before we start looping. then we return the file to store inside list dict.\n",
    "Then list in profile will bes sent to def pattern_probability"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reference=\"https://github.com/andreyrozumnyi/rosalind/blob/master/Chapter%202/BA2D.py\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
