{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAAAAAAGGGAATCA\n",
      "GAGTTCAGGCAAGAA\n",
      "AAGCGAAGGCAAGGA\n",
      "TCGAGAAGGCAAGAG\n",
      "GAGAGATTCCAAGAA\n",
      "GAGTTCAGGCAAGAA\n",
      "GAGAGAATATAAGAA\n",
      "AAGAGAAGGCAAGGT\n",
      "GAGAGATCCCAAGAA\n",
      "GAGAATGGGCAAGAA\n",
      "GAGAGCGTGCAAGAA\n",
      "TGCAGAAGGCAAGAA\n",
      "GAAGCAAGGCAAGAA\n",
      "GAGAGAAGGCAAATC\n",
      "GAGAGAAGGCGGAAA\n",
      "GTTTGAAGGCAAGAA\n",
      "GAGACGGGGCAAGAA\n",
      "GAGAGAAGGCACAGA\n",
      "GAGAGTTAGCAAGAA\n",
      "GAGAGAAGCGGAGAA\n"
     ]
    }
   ],
   "source": [
    "from random import randint, choice\n",
    "import sys\n",
    "import numpy\n",
    "\n",
    "def score(motifs):\n",
    "    score = 0\n",
    "    for i in range(len(motifs[0])):\n",
    "        motif = ''.join([motifs[j][i] for j in range(len(motifs))])\n",
    "        score += min([hamming_distance(motif, homogeneous*len(motif)) for homogeneous in 'ACGT'])\n",
    "    return score\n",
    "\n",
    "def profile_with_pseudocounts(motifs):\n",
    "    prof = []\n",
    "    for i in range(len(motifs[0])):\n",
    "        col = ''.join([motifs[j][i] for j in range(len(motifs))])\n",
    "        prof.append([float(col.count(nuc)+1)/float(len(col)+4) for nuc in 'ACGT'])\n",
    "    return prof\n",
    "\n",
    "def profile_most_probable_kmer(dna, k, prof):\n",
    "    nuc_loc = {nucleotide:index for index,nucleotide in enumerate('ACGT')}\n",
    "    max_prob = [-1, None]\n",
    "    for i in range(len(dna)-k+1):\n",
    "        current_prob = 1\n",
    "        for j, nucleotide in enumerate(dna[i:i+k]):\n",
    "            current_prob *= prof[j][nuc_loc[nucleotide]]\n",
    "        if current_prob > max_prob[0]:\n",
    "            max_prob = [current_prob, dna[i:i+k]]\n",
    "\n",
    "    return max_prob[1]\n",
    "\n",
    "def motifs_from_profile(profile, dna, k):\n",
    "    return [profile_most_probable_kmer(seq,k,profile) for seq in dna]\n",
    "\n",
    "def randomized_motif_search(dna_list,k,t):\n",
    "    rand_ints = [randint(0,len(dna_list[0])-k) for a in range(t)]\n",
    "    motifs = [dna_list[i][r:r+k] for i,r in enumerate(rand_ints)]\n",
    "\n",
    "    # Initialize the best score as a score higher than the highest possible score.\n",
    "    best_score = [score(motifs), motifs]\n",
    "\n",
    "    # Iterate motifs.\n",
    "    while True:\n",
    "        current_profile = profile_with_pseudocounts(motifs)\n",
    "        motifs = motifs_from_profile(current_profile, dna_list, k)\n",
    "        current_score = score(motifs)\n",
    "        if current_score < best_score[0]:\n",
    "            best_score = [current_score, motifs]\n",
    "        else:\n",
    "            return best_score\n",
    "\n",
    "def hamming_distance(str1, str2):\n",
    "    counter = 0\n",
    "    for s1, s2 in zip(str1, str2):\n",
    "        if s1 != s2:\n",
    "            counter += 1\n",
    "    return counter\n",
    "\n",
    "\n",
    "def profile_randomized_kmer(dna, k, prof):\n",
    "    nuc_loc = {nucleotide: index for index, nucleotide in enumerate('ACGT')}\n",
    "    probs = []\n",
    "    for i in range(len(dna) - k):\n",
    "        current_prob = 1.\n",
    "        for j, nucleotide in enumerate(dna[i:i + k]):\n",
    "            current_prob *= prof[j][nuc_loc[nucleotide]]\n",
    "        probs.append(current_prob)\n",
    "\n",
    "    i = numpy.random.choice(len(probs), p = numpy.array(probs) / numpy.sum(probs))\n",
    "    return dna[i:i + k]\n",
    "\n",
    "\n",
    "def gibbs_sampling_motif_search(dna_list, k, t, N, init_motifs=None):\n",
    "    if init_motifs:\n",
    "        motifs = init_motifs\n",
    "    else:\n",
    "        rand_ints = [randint(0, len(dna_list[0]) - k) for a in range(t)]\n",
    "        motifs = [dna_list[i][r:r + k] for i, r in enumerate(rand_ints)]\n",
    "\n",
    "    best_score = [score(motifs), list(motifs)]\n",
    "\n",
    "    for j in range(N):\n",
    "        i = randint(0, t - 1)\n",
    "        current_profile = profile_with_pseudocounts([x for amotif, x in enumerate(motifs) if amotif != i])\n",
    "        motifs[i] = profile_randomized_kmer(dna_list[i], k, current_profile)\n",
    "        current_score = score(motifs)\n",
    "        if current_score < best_score[0]:\n",
    "            best_score = [current_score, list(motifs)]\n",
    "\n",
    "    return best_score\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    data = \"\".join(open('gibbs.txt')).split()\n",
    "    k, t, N = int(data[0]), int(data[1]), int(data[2])\n",
    "    dna_list = data[3:]\n",
    "    best_motifs = [k * t, None]\n",
    "    for repeat in range(20):\n",
    "        current_motifs = gibbs_sampling_motif_search(dna_list, k, t, N)\n",
    "        if current_motifs[0] < best_motifs[0]:\n",
    "            best_motifs = current_motifs\n",
    "    print('\\n'.join(best_motifs[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GibbsSampler is a more cautious iterative algorithm that discards a single k-mer from the current set of motifs at each iteration and decides to either keep it or replace it with a new one. This algorithm thus moves with more caution in the space of all motifs.\n",
    "\n",
    "Like RandomizedMotifSearch, GibbsSampler starts with randomly chosen k-mers in each of t DNA sequences, but it makes a random rather than a deterministic choice at each iteration. It uses randomly selected k-mers (Motif1, …, Motift) to come up with another (hopefully better scoring) set of k-mers. In contrast with RandomizedMotifSearch (which deterministically defines new motifs as\n",
    "\n",
    "Motifs(Profile(Motifs), Dna),\n",
    "GibbsSampler randomly selects an integer i between 1 and t and then randomly changes only a single k-mer Motifi.\n",
    "\n",
    "To describe how GibbsSampler updates Motifs, we will need a slightly more advanced random number generator. Given a probability distribution (p1, …, pn), this random number generator, denoted Random(p1, …, pn), models an n-sided biased die and returns integer i with probability pi. For example, the standard six-sided fair die represents the random number generator Random(1/6, 1/6, 1/6, 1/6, 1/6, 1/6), whereas a biased die might represent the random number generator Random(0.1, 0.2, 0.3, 0.05, 0.1, 0.25). GibbsSampler further generalizes the random number generator by using the function Random(p1, …, pn) defined for any set of non-negative numbers, i.e., not necessarily satisfying the condition that the pi sum to 1. If the pi sum to some C > 0 instead, then Random(p1, …, pn) is defined as Random(p1/C, …, pn/C), where (p1/C, …, pn/C) is a probability distribution. For example, for (0.1, 0.2, 0.3) with 0.1 + 0.2 + 0.3 = 0.6,\n",
    "\n",
    "Random(0.1, 0.2, 0.3) = Random(0.1/0.6, 0.2/0.6, 0.3/0.6) = Random(1/6, 1/3, 1/2).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reference=\"https://github.com/Butskov/Bioinformatics-Algorithms/blob/master/week3-4/9-Implement%20GibbsSampler.py\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
