{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "def iterate_randomized_motif_search(dna, k ,t):\n",
    "    i = 0\n",
    "    best_motifs = []\n",
    "    for i in range(t):\n",
    "        best_motifs.append(dna[i][:k])\n",
    "    while i < 1000:\n",
    "        motifs = randomized_motif_search(dna, k, t)\n",
    "        if score(motifs) < score(best_motifs):\n",
    "            best_motifs = motifs\n",
    "        i = i+1\n",
    "        #print(i)\n",
    "    return best_motifs\n",
    "\n",
    "\n",
    "def randomized_motif_search(dna, k, t):\n",
    "    best_motifs = random_motifs(dna, k)\n",
    "    while True:\n",
    "        profile = create_profile(best_motifs)\n",
    "        motifs = create_motifs(profile, dna)\n",
    "        if score(motifs) < score(best_motifs):\n",
    "            best_motifs = motifs\n",
    "        else:\n",
    "            return best_motifs\n",
    "\n",
    "\n",
    "def create_motifs(profile, dna):\n",
    "    motifs = []\n",
    "    for i in dna:\n",
    "        motifs.append(profile_most_probable(i, profile, len(profile)))\n",
    "    return motifs\n",
    "\n",
    "\n",
    "def calculate_probablity(profile, string):\n",
    "    probablity = 1\n",
    "    for i in range(0, len(string)):\n",
    "        if string[i] == 'A':\n",
    "            probablity = probablity * profile[i][0]\n",
    "        elif string[i] == 'C':\n",
    "            probablity = probablity * profile[i][1]\n",
    "        elif string[i] == 'G':\n",
    "            probablity = probablity * profile[i][2]\n",
    "        elif string[i] == 'T':\n",
    "            probablity = probablity * profile[i][3]\n",
    "    return probablity\n",
    "\n",
    "\n",
    "def profile_most_probable(dna_string, profile, k):\n",
    "    best_pattern = dna_string[0:0 + k]\n",
    "    best_probability = 0\n",
    "    for i in range(len(dna_string) - k + 1):\n",
    "        string = dna_string[i:i + k]\n",
    "        new_probablity = calculate_probablity(profile, string)\n",
    "        if new_probablity > best_probability:\n",
    "            best_pattern = string\n",
    "            best_probability = new_probablity\n",
    "    return best_pattern\n",
    "\n",
    "\n",
    "def create_profile(motifs):\n",
    "    profile =[]\n",
    "    for i in range(len(motifs[0])):\n",
    "        count_A, count_C, count_G, count_T = 1, 1, 1, 1\n",
    "        for motif in motifs:\n",
    "            if motif[i] == 'A':\n",
    "                count_A += 1\n",
    "            elif motif[i] == 'C':\n",
    "                count_C += 1\n",
    "            elif motif[i] == 'G':\n",
    "                count_G += 1\n",
    "            elif motif[i] == 'T':\n",
    "                count_T += 1\n",
    "        profile.append([count_A / (len(motifs) + 4), count_C/ (len(motifs) + 4),\n",
    "                        count_G / (len(motifs) + 4), count_T / (len(motifs) + 4)])\n",
    "    return profile\n",
    "\n",
    "\n",
    "def hamming_distance(str1, str2):\n",
    "    counter = 0\n",
    "    for s1, s2 in zip(str1, str2):\n",
    "        if s1 != s2:\n",
    "            counter += 1\n",
    "    return counter\n",
    "\n",
    "\n",
    "def find_consensus(motifs):\n",
    "    consensus = ''\n",
    "    for i in range(len(motifs[0])):\n",
    "        count_A, count_C, count_G, count_T = 0, 0, 0, 0\n",
    "        for motif in motifs:\n",
    "            if motif[i] == 'A':\n",
    "                count_A += 1\n",
    "            elif motif[i] == 'C':\n",
    "                count_C += 1\n",
    "            elif motif[i] == 'G':\n",
    "                count_G += 1\n",
    "            elif motif[i] == 'T':\n",
    "                count_T += 1\n",
    "        if count_A >= max(count_C, count_G, count_T):\n",
    "            consensus += \"A\"\n",
    "        elif count_C >= max(count_A, count_G, count_T):\n",
    "            consensus += \"C\"\n",
    "        elif count_G >= max(count_C, count_A, count_T):\n",
    "            consensus += \"G\"\n",
    "        elif count_T >= max(count_C, count_G, count_A):\n",
    "            consensus += \"T\"\n",
    "    return consensus\n",
    "\n",
    "\n",
    "def score(motifs):\n",
    "    consensus = find_consensus(motifs)\n",
    "    score = 0\n",
    "    for motif in motifs:\n",
    "        score += hamming_distance(consensus, motif)\n",
    "    return score\n",
    "\n",
    "\n",
    "def random_motifs(dna,k):\n",
    "    motifs = []\n",
    "    for i in dna:\n",
    "        point = random.randint(0, len(i)-k)\n",
    "        motifs.append(i[point:point + k])\n",
    "    return motifs\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    data = \"\".join(open('randomized_motif_search.txt')).split()\n",
    "    best_motifs = iterate_randomized_motif_search(data[2:], int(data[0]), int(data[1]))\n",
    "    for i in best_motifs:\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Randomized algorithms may be non-intuitive because they lack the control of traditional algorithms. Some randomized algorithms are Las Vegas algorithms, which deliver solutions that are guaranteed to be exact, despite the fact that they rely on making random decisions. Yet most randomized algorithms are Monte Carlo algorithms. These algorithms are not guaranteed to return exact solutions, but they do quickly find approximate solutions. Because of their speed, they can be run many times, allowing us to choose the best approximation from thousands of runs.\n",
    "\n",
    "Exactly what RandomizedMotifSearch does. To implement this algorithm, you will need to randomly select the initial collection of k-mers that form the motif matrix Motifs. To do so, you will need a random number generator (denoted RandomNumber(N)) that is equally likely to return any integer from 1 to N. You might like to think about this random number generator as an unbiased N-sided die.\n",
    "\n",
    "What diffrent in this code from previous are we improve the code by adding import random function and at the main function we have iterate function to make it iterate randomly like a dice."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reference=\"https://github.com/Butskov/Bioinformatics-Algorithms/blob/master/week3-4/8-Implement%20RandomizedMotifSearch.py\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
